### DISCUSSION TITLE
Future command enhancements

### DISCUSSION THREAD

**Discussion by dnewman-gpsw** [gopro/labs#210] [General]
While Labs is constantly adding hacks to help you create the perfect image, much of the core command set hasn't changed since the first introduction on HERO8.  In the next Labs release we hope to expand the command language to have more if-this-then-that conditional commands, as there are some great ideas that may be achievable with these new scripting controls.  

Today's conditionals are only time based.  If(time > value) do_command_A else do_command_B. e.g. Labs command `>10:00mV~mT`.  This is good for day versus night controls, but little more.  Coming soon are other camera conditionals. The command format is <xValue or >xValue  where x is a status type, and Value is used in the conditionals that <xValue is the same as if(x_status < Value) then ...

Some status type:
- a acceleration - >aValueCMD if(acceleration > Value) then CMD, units in Gs
- b battery - >bValueCMD if(battery > Value) then CMD, units in percentage
- c Coord Distance - >cDistCMD  if(distance > Value) then CMD, units in meters, compare distance from initial GPS location
- c:X CoordDistance - >c:XdistCMD There are up to 26 pre-stored GPS locations oMFRMA=latt,long thru oMFRMZ. If nothing is store in FRMx, it will initialize with the current location.
- d GPS DOP - <dValueCMD - if(DOP < Value) then CMD, units in DOP. GPS location precision.
- e random - <eValueCMD <e50 - 50% true <e90 - 90% true.
- g gyro  - >gValueCMD if(gryo > Value) CMD, numbers are in degrees per second.
- h height - >hValueCMD if(height > Value) CMD, numbers are in meters above sealevel.
- k speed - >kValueCMD if(gps Speed > Value) CMD e.g. >k45!S, numbers are in km/h.
- p sound pressure level - >pValueCMD if(spl > Value) CMD, numbers are in dB
- i iso - >iValueCMD - testing ISO where ISO ranges from 100 to 6400
- r recording - >rCMD1~CMD2 if(Recording) then CMD1 else CMD2
- r:C remote Connected - >r:CCMD1~CMD2 if(RC_Connected) then CMD1 else CMD2 
- r:A remote App Connected - >r:ACMD1~CMD2 if(App_Connected) then CMD1 else CMD2 
- s shutter - >sValueCMD - testing shutter, where 1/Value is used for shutter speed
- t:X timedate - >t:XValueCMD - where X, Y-Year M-Month D-Day H-Hour N-miNute S-second W-day_of_the_Week B-seconds_since_Boot Q-seconds_since_Qrcode
- u USB power - >uCMD1~CMD2 if(power is on USB) then CMD1 else CMD2

With these command you can create custom capture motion triggers, or geofencing a capture, setting a region of interest for capture to auto start.

oMFRMA=36.5854815,-121.7640418!D<r>cA:1000!S~"%dm to Laguna Seca"+!600R
This will start recording when within 1000 meters of Laguna Seca raceway.

Explained:
`oMFRMA=36.5854815,-121.7640418` -  sets the GPS coordinate of capture region
`!D` - waits for a GPS lock
`<r` - if not recording yet
`  <c:A1000` - if distance to coordinate A is less than 1000 meters
`      !S` - start recording
`  ~` - else
`      "%dm to Laguna Seca"`  - display the distance the location
`      +` - more in else
`      !600R` - Sleep for 10 minutes and repeat

The above uses only the recording status and the coordinate distance status.

Other examples:
- Sky drivers could create scripts to start capture when height above sea level is > 3000 meters 
- Dashcam script could record only while the car is motion, and charge when not, while the car is on.

Please suggest other status variables to look for (only used 13 characters so far.)

PS. I thought of one more - day of the week, could be useful is work week construction time-lapses.

PPS. These features could be a month or so away, so this is a good time for feedback.

**Comment by andrerav** [comment]
Man. This is insane. Nice job! Here's a quick suggestion off the top of my head:

- l lightmeter - >lValueCMD (if lightmeter > Value) then CMD

For scenarios where you want to start/stop recording under certain lighting conditions. Should probably be combined with r and/or sleeping for a simple hysteresis mechanism. No idea what the unit would be, but if this is possible to do then you probably already have some convertable unit of measurement for this in the GoPro.

**Comment by andrerav** [comment]
Temperature sensors readouts (in particular, sensors involved in triggering overheating shutdowns) might be useful. The goal being the ability to stop recording and sleep for a while to cool down before resuming recording to prevent shutdowns due to overheating.

**Comment by dnewman-gpsw** [comment]
Thanks for chiming in, and more is welcome.  You can use the current ISO and shutter values to compute a simple light meter, although there might be a system light level that could be exposed through these controls.  DSP and/or battery temperature could be exposed, although they have less creative applications.  

Also coming are variables and basic math, so lots creative capture solutions will become possible.

**Comment by jxqqk** [comment]
Even if it's mostly an outdoor camera where the GPS is available, for indoor use will be useful to be able to sync the time by connecting to internet, by a time server or even the Gopro cloud service; I noticed that when powered only by usb for few hours a day the internal time shift a lot.
Thank you.
JM

**Reply by dnewman-gpsw** [reply]
While this is a good idea, this is not a typical Labs hack.  The camera has a RTOS (real time operating system) for all the core camera functions, and Linux image that runs all the internet stuff.  Labs runs on the RTOS, so anything to like getting network time over WiFi is simply not available.  For develops doing camera control via BLE or WIFi or USB consider using https://gopro.github.io/OpenGoPro/
